src/core/main.c

```diff
@@ -2004,15 +2004,16 @@ static void log_execution_mode(bool *ret_first_boot) {
                         *ret_first_boot = false;
                         log_info("Running in initial RAM disk.");
                 } else {
-                        /* Let's check whether we are in first boot, i.e. whether /etc is still unpopulated. We use
-                         * /etc/machine-id as flag file, for this: if it exists we assume /etc is populated, if it
-                         * doesn't it's unpopulated. This allows container managers and installers to provision a
-                         * couple of files already. If the container manager wants to provision the machine ID itself
-                         * it should pass $container_uuid to PID 1. */
-
-                        *ret_first_boot = access("/etc/machine-id", F_OK) < 0;
+                        /* Let's check whether we are in first boot, i.e. whether /etc is still unpopulated.
+                         * We use /etc/machine-id as flag file, for this: if it exists and is not empty we
+                         * assume /etc is populated, if it doesn't it's unpopulated. This allows container
+                         * managers and installers to provision a couple of files already. If the container
+                         * manager wants to provision the machine ID itself it should pass $container_uuid to
+                         * PID 1. */
+
+                        *ret_first_boot = !!missing_or_empty_path("/etc/machine-id");
```

Comment:
Yeah, you were actually too fast for me to write a comment about it:  I think my current implementation is broken because of this.  Apparently most distros ship an empty /etc/machine-id which leads to systemd (with my changes) presetting all units.  That has caused the integration tests in CI to fail and would probably cause a lot of problems further down the line (from reading #8268).

So my idea was exactly what you describe: A 'uninitialized' value for the machine-id until the first boot succeeds.

---

That said, I think the CI failures might have made an unrelated problem surface:  Currently `proc-sys-fs-binfmt_misc.mount` would be preset to enabled which doesn't seem in line with its documentation and, as I observed, leads to various issues.  The temporary commit (e25743ff423bb2e98228a6e117e1a74f8c3515e3) I added seems to fix it.  What do you think?  Maybe I should open a separate PR for this?

Reply:
Hmm, sounds OK to include it here in a separate commit. (Also see 508133917d9628f8c0adc5db34d595196b72b1c8)

And yeah, I think it should be disabled by default. /cc @keszybz

Reply:
> Hmm, sounds OK to include it here in a separate commit. (Also see 5081339)

+1



> * if /etc/machine-id exists and is initialized → all good, we do nothing
> 
> * if /etc/machine-id is missing → we initialize it and declare first boot mode
> 
> * if /etc/machine-id exists and is empty → we initialize/overmount it, but do not enter first boot mode
> 
> * if /etc/machine-id exists and contains special string "uninitialized" → we initialize/overmount it and declare first boot mode

Shouldn't the last two be reversed for backwards compatibility? IIUC, currently we enter first boot mode when /etc/machine-id exists and is empty. Instead of redefining this, I'd define a special value that means "use a temporary value on each boot". Maybe even `0000000000000000`.

Reply:
@keszybz, if /etc/machine-id exists but is empty, this is currently **not** considered a first boot and the idea was to keep this as is (for backwards compatibility).  So there is no redefinition of any exisiting value here.